pkg/status/combinedstatus-resolver.go

```diff
@@ -0,0 +1,472 @@
+// that should be queued for syncing.
+func (c *combinedStatusResolver) NoteBindingResolution(bindingName string, bindingResolution binding.Resolution,
+	deleted bool, workStatuses []*workStatus,
+	statusCollectorLister controllisters.StatusCollectorLister) sets.Set[util.ObjectIdentifier] {
+	c.Lock()
+	defer c.Unlock()
+
+	combinedStatusIdentifiersToQueue := sets.New[util.ObjectIdentifier]()
+	workloadIdentifiersToEvaluate := sets.New[util.ObjectIdentifier]()
+	// destinations set to collect from, shared for all workload objects
+	collectionDestinations := sets.New(abstract.SliceMap(bindingResolution.Destinations,
+		func(destination v1alpha1.Destination) string { return destination.ClusterId })...)
+
+	// (1)
+	if deleted {
+		// for every combinedstatus resolution - queue its object identifier for syncing
+		// and remove the resolution from memory
+		for workloadObjectIdentifier, _ := range c.bindingNameToResolutions[bindingName] {
+			combinedStatusIdentifiersToQueue.Insert(getCombinedStatusIdentifier(bindingName, workloadObjectIdentifier))
+		}
+
+		delete(c.bindingNameToResolutions, bindingName)
+		return combinedStatusIdentifiersToQueue
+	}
+
+	// if the binding resolution is not yet noted - create a new entry
+	objectIdentifierToResolution, exists := c.bindingNameToResolutions[bindingName]
+	if !exists {
+		objectIdentifierToResolution = make(map[util.ObjectIdentifier]*combinedStatusResolution)
+		c.bindingNameToResolutions[bindingName] = objectIdentifierToResolution
+	}
+
+	// (2) remove excessive combinedstatus resolutions of objects that are no longer
+	// associated with the binding resolution
+	for objectIdentifier, _ := range objectIdentifierToResolution {
+		if _, exists := bindingResolution.ObjectIdentifierToData[objectIdentifier]; !exists {
+			combinedStatusIdentifiersToQueue.Insert(getCombinedStatusIdentifier(bindingName, objectIdentifier))
+			delete(objectIdentifierToResolution, objectIdentifier)
+		}
+	}
+
+	// (~2+3) create/update combinedstatus resolutions for every object that requires status collection,
+	// and delete resolutions that are no longer required
+	for objectIdentifier, objectData := range bindingResolution.ObjectIdentifierToData {
+		csResolution, exists := objectIdentifierToResolution[objectIdentifier]
+		if len(objectData.StatusCollectors) == 0 {
+			if exists { // associated resolution is no longer required
+				combinedStatusIdentifiersToQueue.Insert(getCombinedStatusIdentifier(bindingName, objectIdentifier))
+				delete(objectIdentifierToResolution, objectIdentifier)
+			}
+
+			continue
+		}
+
+		// create resolution entry if missing
+		if !exists {
+			csResolution = &combinedStatusResolution{
+				statusCollectorNameToData: make(map[string]*statusCollectorData),
+			}
+			objectIdentifierToResolution[objectIdentifier] = csResolution
+		}
+
+		// fetch missing statuscollector specs
+		c.fetchMissingStatusCollectorSpecsLocked(statusCollectorLister, objectData.StatusCollectors)
```

Comment:
FYI, the method naming convention should be explained up front somewhere.

Reply:
Will be addressed in followup PR - logged in issue #2141.